pkg/core/transport_test.go

```diff
@@ -0,0 +1,248 @@
+				"src": addr,
+			}).Info("Client connected")
+		}
+		return ok, msg
+	}
+
+	server := &HysteriaTransport{
+		addr:              server_addr,
+		protocol:          protocol,
+		tlsConfig:         serverTlsConfig,
+		quicConfig:        quicConfig,
+		transport:         transport.DefaultServerTransport,
+		sendBPS:           0,
+		recvBPS:           0,
+		congestionFactory: congestionFactory,
+		disableUDP:        false,
+		obfuscator:        obfuscator,
+		connectFunc:       connectFunc,
+		disconnectFunc:    disconnectFunc,
+	}
+
+	l, err := server.Listen()
+
+	if err != nil {
+		fmt.Println("Failed to initialize server")
+	}
+
+	fmt.Println("Server up and running")
+
+	serverConn, err := l.Accept()
+	defer serverConn.Close()
+
+	if err != nil {
+		return err
+	}
+
+	serverBuffer := make([]byte, len(test_request))
+	serverConn.SetReadDeadline(time.Now().Add(time.Second * 20))
+	_, err = serverConn.Read(serverBuffer)
+
+	if err != nil {
+		return err
+	}
+
+	s := string(serverBuffer)
+	if s == test_request {
+		serverConn.SetWriteDeadline(time.Now().Add(time.Second * 2))
+		serverConn.Write([]byte(test_response))
+		serverConn.Close()
+		return nil
+	}
+
+	return errors.New("Something is wrong")
+}
+
+// Simulate a client
+func runClient(obfuscator *obfs.XPlusObfuscator) error {
+	// Load TLS client config
+	var clientTlsConfig = &tls.Config{
+		InsecureSkipVerify: false,
+		MinVersion:         tls.VersionTLS13,
+		NextProtos:         []string{DefaultALPN},
+		ServerName:         server_name,
+	}
+	bs, err := ioutil.ReadFile(customCA)
+	if err != nil {
+		logrus.WithFields(logrus.Fields{
+			"error": err,
+			"file":  customCA,
+		}).Fatal("Failed to load CA")
+	}
+	cp := x509.NewCertPool()
+	if !cp.AppendCertsFromPEM(bs) {
+		logrus.WithFields(logrus.Fields{
+			"file": customCA,
+		}).Fatal("Failed to parse CA")
+	}
+	clientTlsConfig.RootCAs = cp
+
+	// QUIC config
+	quicConfig := &quic.Config{
+		InitialStreamReceiveWindow:     DefaultStreamReceiveWindow,
+		MaxStreamReceiveWindow:         DefaultStreamReceiveWindow,
+		InitialConnectionReceiveWindow: DefaultConnectionReceiveWindow,
+		MaxConnectionReceiveWindow:     DefaultConnectionReceiveWindow,
+		KeepAlive:                      true,
+		DisablePathMTUDiscovery:        false,
+		EnableDatagrams:                true,
+	}
+
+	client, err := NewClient(server_addr, protocol, []byte(auth_str), clientTlsConfig, quicConfig,
+		transport.DefaultClientTransport, client_up_mbps, client_down_mbps,
+		congestionFactory, obfuscator)
+
+	if err != nil {
+		fmt.Println("Failed to initialize client")
+		return err
+	}
+
+	fmt.Println("Client up and running")
+	clientConn, err := client.Dial()
+	defer clientConn.Close()
+
+	if err != nil {
+		fmt.Println("Failed to connect to the server")
+		return err
+	}
+
+	// write data from clientConn for server to read
+	clientConn.SetWriteDeadline(time.Now().Add(time.Second * 2))
+	_, err = clientConn.Write([]byte(test_request))
```

Comment:
I am stuck at finishing up this E2E test at the moment. I have gotten client and server handshake working, but in this line: the client sends a simple message to the server, but the server read result is empty.

```
Server up and running
time="2022-05-29T16:52:38Z" level=info msg="Client connected" src="127.0.0.1:40141"
Client up and running
```

Once I am unblocked here, I will write a utility to generate certs instead of the ones I hard-coded at the beginning of the test.

Reply:
> I will write a utility to generate certs instead of the ones I hard-coded at the beginning of the test.

Found this    

https://github.com/lucas-clemente/quic-go/blob/b5ef99a32c250fc63f89cc686c13a008c5419d01/example/echo/echo.go#L92-L113